### ISSUE TITLE
cannot find -lcudart

### ISSUE THREAD

**Issue by ar4** [ar4/deepwave#51]
> Hi, Alan.
> Thanks for the deepwave.
> I also have some problem installing the newest version of deepwave. I used the command " conda install pytorch==1.12.0 torchvision==0.13.0 torchaudio==0.12.0 cudatoolkit=10.2 -c pytorch" to install pytorch and then used "pip install deepwave" . But when I tried to "import deep wave", I have the error below. **So could you please tell me which version of pytorch you used?**:
> 
> Traceback (most recent call last):
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1808, in _run_ninja_build
>     subprocess.run(
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/subprocess.py", line 516, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/nas/home/wxu/lsrtmdip/lsrtm_test_Marmousi.py", line 4, in <module>
>     import deepwave
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/deepwave/__init__.py", line 26, in <module>
>     torch.utils.cpp_extension.load(
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1202, in load
>     return _jit_compile(
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1425, in _jit_compile
>     _write_ninja_file_and_build_library(
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1537, in _write_ninja_file_and_build_library
>     _run_ninja_build(
>   File "/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1824, in _run_ninja_build
>     raise RuntimeError(message) from e
> RuntimeError: Error building extension 'deepwave': [1/5] c++ -MMD -MF scalar.o.d -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -march=native -Ofast -fopenmp -c /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/deepwave/scalar.cpp -o scalar.o 
> [2/5] c++ -MMD -MF scalar_born.o.d -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -march=native -Ofast -fopenmp -c /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/deepwave/scalar_born.cpp -o scalar_born.o 
> [3/5] /nas/home/wxu/miniconda3/envs/lsrtmdip/bin/nvcc  -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_75,code=compute_75 -gencode=arch=compute_75,code=sm_75 --compiler-options '-fPIC' --restrict -O3 --use_fast_math -std=c++14 -c /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/deepwave/scalar.cu -o scalar.cuda.o 
> [4/5] /nas/home/wxu/miniconda3/envs/lsrtmdip/bin/nvcc  -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdip/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_75,code=compute_75 -gencode=arch=compute_75,code=sm_75 --compiler-options '-fPIC' --restrict -O3 --use_fast_math -std=c++14 -c /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/deepwave/scalar_born.cu -o scalar_born.cuda.o 
> [5/5] c++ scalar.o scalar_born.o scalar.cuda.o scalar_born.cuda.o -shared -march=native -Ofast -fopenmp -L/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda_cu -ltorch_cuda_cpp -ltorch -ltorch_python -L/nas/home/wxu/miniconda3/envs/lsrtmdip/lib64 -lcudart -o deepwave.so
> FAILED: deepwave.so 
> c++ scalar.o scalar_born.o scalar.cuda.o scalar_born.cuda.o -shared -march=native -Ofast -fopenmp -L/nas/home/wxu/miniconda3/envs/lsrtmdip/lib/python3.8/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda_cu -ltorch_cuda_cpp -ltorch -ltorch_python -L/nas/home/wxu/miniconda3/envs/lsrtmdip/lib64 -lcudart -o deepwave.so
> /usr/bin/ld: cannot find -lcudart: No such file or directory
> collect2: error: ld returned 1 exit status
> ninja: build stopped: subcommand failed.
> 
> 
> Process finished with exit code 1

_Originally posted by @XuVV in https://github.com/ar4/deepwave/issues/50#issuecomment-1441815418_

Hello @XuVV. I am moving your issue here as I think it has a different cause than Naveen's.

I suspect that NVCC is looking for the CUDA runtime library (libcudart) in the wrong place. To check, please report the output of these commands:
ls /home/wxu/miniconda3/envs/lsrtmdip/lib64/libcudart*
ls /home/wxu/miniconda3/envs/lsrtmdip/lib/libcudart*

**Comment by XuVV** [comment]
Hello, thanks, Alan. This is the output :
ls: cannot access '/home/wxu/miniconda3/envs/lsrtmdip/lib64/libcudart*': No such file or directory
ls: cannot access '/home/wxu/miniconda3/envs/lsrtmdip/lib/libcudart*': No such file or directory

**Comment by XuVV** [comment]
Hello, Alan. 
When I used this version of pytorch "conda install pytorch==1.7.1 torchvision==0.8.2 torchaudio==0.7.2 cudatoolkit=11.0 -c pytorch" and this version of deepwave "pip install deepwave==0.0.8", it worked.
So could you please tell me which version of pytorch do you use when you use the newest version of deepwave?

**Comment by ar4** [comment]
Hmm. The CUDA library doesn't seem to be where it expects it to be. Can you try running "echo $CUDA_HOME" when you are in the Conda environment you are trying to install Deepwave in, to see what that environment variable is set to?

Is there a reason why you are trying to use an old version of CUDA? If you just use the current PyTorch install command ("conda install pytorch torchvision torchaudio pytorch-cuda=11.6 -c pytorch -c nvidia") in a new Conda environment, and then try to install Deepwave, does it work?

I am currently using PyTorch 1.12.1.

**Comment by XuVV** [comment]
Hello, Alan.
When I tried to run "echo $CUDA_HOME", it outputs nothing. I don't know why.
There is no special reason I used an old version of CUDA, just followed someone else.
When I try to  use the current PyTorch install command ("conda install pytorch torchvision torchaudio pytorch-cuda=11.6 -c pytorch -c nvidia") in a new Conda environment, and then try to install Deepwave, it doesn't work still, which has the error:
Traceback (most recent call last):
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1900, in _run_ninja_build
    subprocess.run(
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nas/home/wxu/lsrtmdip/lsrtm_test_Marmousi.py", line 4, in <module>
    import deepwave
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/__init__.py", line 31, in <module>
    torch.utils.cpp_extension.load(
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1284, in load
    return _jit_compile(
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1508, in _jit_compile
    _write_ninja_file_and_build_library(
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1623, in _write_ninja_file_and_build_library
    _run_ninja_build(
  File "/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1916, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error building extension 'deepwave': [1/7] /nas/home/wxu/miniconda3/envs/lsrtmdipnew/bin/nvcc  -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_75,code=compute_75 -gencode=arch=compute_75,code=sm_75 --compiler-options '-fPIC' --restrict -O3 --use_fast_math -std=c++14 -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/scalar.cu -o scalar.cuda.o 
[2/7] /nas/home/wxu/miniconda3/envs/lsrtmdipnew/bin/nvcc  -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_75,code=compute_75 -gencode=arch=compute_75,code=sm_75 --compiler-options '-fPIC' --restrict -O3 --use_fast_math -std=c++14 -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/elastic.cu -o elastic.cuda.o 
/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/elastic.cu(1613): warning #177-D: function "<unnamed>::zero_edge_bottom" was declared but never referenced

/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/elastic.cu(1618): warning #177-D: function "<unnamed>::zero_edge_left" was declared but never referenced

[3/7] /nas/home/wxu/miniconda3/envs/lsrtmdipnew/bin/nvcc  -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_75,code=compute_75 -gencode=arch=compute_75,code=sm_75 --compiler-options '-fPIC' --restrict -O3 --use_fast_math -std=c++14 -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/scalar_born.cu -o scalar_born.cuda.o 
[4/7] c++ -MMD -MF scalar.o.d -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -march=native -Ofast -fopenmp -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/scalar.cpp -o scalar.o 
[5/7] c++ -MMD -MF elastic.o.d -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -march=native -Ofast -fopenmp -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/elastic.cpp -o elastic.o 
[6/7] c++ -MMD -MF scalar_born.o.d -DTORCH_EXTENSION_NAME=deepwave -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/TH -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/include/THC -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include -isystem /nas/home/wxu/miniconda3/envs/lsrtmdipnew/include/python3.8 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -march=native -Ofast -fopenmp -c /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/deepwave/scalar_born.cpp -o scalar_born.o 
[7/7] c++ scalar.o scalar_born.o elastic.o scalar.cuda.o scalar_born.cuda.o elastic.cuda.o -shared -march=native -Ofast -fopenmp -L/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda_cu -ltorch_cuda_cpp -ltorch -ltorch_python -L/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib64 -lcudart -o deepwave.so
FAILED: deepwave.so 
c++ scalar.o scalar_born.o elastic.o scalar.cuda.o scalar_born.cuda.o elastic.cuda.o -shared -march=native -Ofast -fopenmp -L/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib/python3.8/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda_cu -ltorch_cuda_cpp -ltorch -ltorch_python -L/nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib64 -lcudart -o deepwave.so
/usr/bin/ld: cannot find -lcudart: No such file or directory
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Process finished with exit code 1

**Comment by XuVV** [comment]
yes, when I runed this python -c "from torch.utils.cpp_extension import CUDA_HOME; print(CUDA_HOME)"
The output is "/nas/home/wxu/miniconda3/envs/lsrtmdipnew".

**Comment by XuVV** [comment]
when I did "export CUDA_HOME=/nas/home/wxu/miniconda3/envs/lsrtmdipnew", and then tried to import deepwave, there is still the same problem.

**Comment by ar4** [comment]
OK. Can you try to find libcudart.so in that directory ("/nas/home/wxu/miniconda3/envs/lsrtmdipnew")? It's usually in the "lib64" subdirectory of it, but compiling Deepwave is failing because apparently it is not there. I have seen instances before where it is instead in the "lib" subdirectory, but when we searched before (in a different environment, but the result will probably be the same), it wasn't there either. Going to that directory and running the command `find . -name "libcudart*"` might help to locate it.

Also, just to confirm: you have an appropriate GPU, and didn't get any error messages when installing PyTorch?

**Comment by ar4** [comment]
OK, so in this environment there is a libcudart.so in the "lib" directory. That is promising! I think you will thus solve your problem if you create a softlink from "lib64" to "lib". You can do that with this command: `ln -s /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib /nas/home/wxu/miniconda3/envs/lsrtmdipnew/lib64"` (or, if it is easier, go to /nas/home/wxu/miniconda3/envs/lsrtmdipnew and then run `ln -s lib lib64`). You can then try importing Deepwave again, and hopefully it will work this time.

**Comment by XuVV** [comment]
Thank you very much, Doctor. It WORKED!
But I still have a stupied question: do I set the "export CUDA_HOME=/nas/home/wxu/miniconda3/envs/lsrtmdipnew" wrong?
How can I use the cuda in the conda environment?

**Comment by ar4** [comment]
That is good news! Thank you for your patience. Hopefully you will find Deepwave useful enough to be worth the effort.

I see now that we could have solved this much earlier - I made a mistake in the first command I asked you to run. I wrote `ls /home/wxu/miniconda3/envs/lsrtmdip/lib/libcudart*`, but I should have written `ls /nas/home/wxu/miniconda3/envs/lsrtmdip/lib/libcudart*`. I'm sorry!

There is no need for you to set CUDA_HOME. I don't think the problem was caused by you. For some reason the CUDA libraries are sometimes installed in the "lib" directory, but PyTorch is hardcoded to expect them in the "lib64" directory. I created an Issue in the PyTorch repository about it a few months ago as this same problem occurred with someone else, but it seems that it has not been fixed yet. I think your CUDA should work correctly - we just created a link from the "lib64" directory that PyTorch expects to the "lib" one where the libraries are actually installed, so PyTorch is able to find them.
